# Homework #1 - The Linear Model

(This skeleton file is provided for HW#1 only.  You are expected to modify it for use in this and other homeworks for the course.)    

This is the README section for A0000000X's submission.
(For group submissions [when applicable], simply concatenate the student matric numbers in lexicographical order separated by a '-' (dash); e.g., A0000000X-A0000001Y)

### General Notes about this assignment 

Place your comments or requests here for the CS3244 staff to read.  Discuss your architecture or experiments in general.  A paragraph or two is usually sufficient.

### Files included with this submission

List the files in your submission here and provide a short 1 line description of each file.  Make sure your submission's files are named and formatted correctly.


## Programming Exercise 1

In [31]:
import numpy as np
from numpy import *  
import numpy.random as nr
import matplotlib.pyplot as plt
%matplotlib inline
import time  
# Plotting with style! 
import seaborn as sb 

# Size the plot appropriately for online display
plt.rcParams['figure.figsize'] = (12.0, 10.0)

Let's fix the random number generator first, in case we need results that are replicable.

In [32]:
nr.seed(3244)

In [33]:
#    Write your solution to the programming assignment here.  We've suggested some cells that you can add 
#    to your notebook as single line comments below.
#    Please place all of your cells to be run in a linear, unintervened order, such that we can automate
#    the running and grading of the assignment.

# load datasets code
# calculate the sigmoid function  
# calculate the sigmoid function  
def sigmoid(inX):  
    return 1.0 / (1 + exp(-inX))  
  # 
def loadData(path):  
    train_x = []  
    train_y = []  
    fileIn = open(path)  
    for line in fileIn.readlines():  
        lineArr = line.strip().split()  

        data = [1.0]

        for i in range(0,20):
        # for i in range(1,3):
            data.append(float(lineArr[i]))

        y = int(lineArr[20])

        if y < 0:
            y = 0

        train_x.append(data)  
        train_y.append(y) 

    # print train_y 
    return mat(train_x), mat(train_y).transpose()  
  
# train a logistic regression model using some optional optimize algorithm  
# input: train_x is a mat datatype, each row stands for one sample  
#        train_y is mat datatype too, each row is the corresponding label  
#        opts is optimize option include step and maximum number of iterations  
def trainLogRegres(train_x, train_y, opts):  
    # calculate training time  
    startTime = time.time()  
  
    numSamples, numFeatures = shape(train_x)  
    alpha = opts['alpha']; maxIter = opts['maxIter']  
    weights = ones((numFeatures, 1))  

    # optimize through gradient descent algorilthm  
    for k in range(maxIter):  
        if opts['optimizeType'] == 'gradDescent': # deterministic gradient descent algorilthm  
            output = sigmoid(train_x * weights)  
            error = train_y - output  
            weights = weights + alpha * train_x.transpose() * error  
        elif opts['optimizeType'] == 'stocGradDescent': # stochastic gradient descent  
            for i in range(numSamples):  
                output = sigmoid(train_x[i, :] * weights)  # pick some of the data
                error = train_y[i, 0] - output  # only considers the error
                weights = weights + alpha * train_x[i, :].transpose() * error  
        else:  
            raise NameError('Not support optimize method type!')  
    
    ein = 0
    
    for i in range(numSamples):
        ein += np.log(1 + exp(- train_x[i, :] * weights * train_y[i , 0]))
    
    ein = ein / numSamples
    
    print('Error of input is: ' + str(ein) + '\n')
    
    return weights  
  
  
# test your trained Logistic Regression model given test set  
def testLogRegres(weights, test_x, test_y):  
    numSamples, numFeatures = shape(test_x)  
    matchCount = 0  
    
    eout = 0
    
    for i in range(numSamples):  
        result = (sigmoid(test_x[i, :] * weights)[0, 0])
        eout += np.log(1 + exp(- test_x[i, :] * weights * test_y[i , 0]))
        predict = result > 0.5

        if predict == (int(test_y[i, 0]) > 0):  
            matchCount += 1  
        #     print str(i) + 'th success: ' + str(result) + '\n'
        # else:
        #     print str(i) + 'th fails: ' + str(result) + '\n'
        
    accuracy = float(matchCount) / numSamples  
    eout = eout / numSamples
    
    print('Error of output is: ' + str(eout) + '\n')
    
    return accuracy  

# show your trained logistic regression model only available with 2-D data  
def showLogRegres(weights, train_x, train_y):  
    # notice: train_x and train_y is mat datatype  
    numSamples, numFeatures = shape(train_x)  
    if numFeatures != 3:  
        print("Sorry! I can not draw because the dimension of your data is not 2!")  
        return 1  
  
    # draw all samples  
    for i in range(numSamples):  
        if int(train_y[i, 0]) == 0:  
            plt.plot(train_x[i, 1], train_x[i, 2], 'or')  
        elif int(train_y[i, 0]) == 1:  
            plt.plot(train_x[i, 1], train_x[i, 2], 'ob')  
  
    # draw the classify line  
    min_x = min(train_x[:, 1])[0, 0]  
    max_x = max(train_x[:, 1])[0, 0]  
    weights = weights.getA()  # convert mat to array  
    y_min_x = float(-weights[0] - weights[1] * min_x) / weights[2]  
    y_max_x = float(-weights[0] - weights[1] * max_x) / weights[2]  
    plt.plot([min_x, max_x], [y_min_x, y_max_x], '-g')  
    plt.xlabel('X1'); plt.ylabel('X2')  
    plt.show()  

## step 1: load data  
print("step 1: load data...")  
train_x, train_y = loadData('hw1-train.dat')  
test_x , test_y = loadData('hw1-test.dat')
  
## step 2: training...  
print("step 2: training...")  
opts = {'alpha': 0.05, 'maxIter': 2333, 'optimizeType': 'stocGradDescent'}  
optimalWeights = trainLogRegres(train_x, train_y, opts)
  
print(optimalWeights.transpose())
## step 3: testing  
print("step 3: testing...")  
accuracy = testLogRegres(optimalWeights, test_x, test_y)  
  
## step 4: show the result  
print("step 4: show the result...")    
print('The classify accuracy is: %.3f%%' % (accuracy * 100))  

# showLogRegres(optimalWeights, train_x, train_y) 
# showLogRegres(optimalWeights, test_x, test_y)    

step 1: load data...
step 2: training...
Error of input is: [[ 0.55564664]]

[[-0.67861174 -0.91886456  1.31188782 -1.62156996  0.19618585 -1.49043856
  -0.0813952   1.89044304 -1.09839404  0.63017345  2.13711557  0.44252764
  -1.43972683 -0.95257355  1.30120275  1.73754254  1.94029756 -1.88386902
   2.03535238 -0.90499649 -1.54060159]]
step 3: testing...
Error of output is: [[ 0.58427082]]

step 4: show the result...
The classify accuracy is: 81.433%


In [34]:
## step 2: training...  
print("step 2: training...")  
opts = {'alpha': 0.005, 'maxIter': 2333, 'optimizeType': 'stocGradDescent'}  
optimalWeights = trainLogRegres(train_x, train_y, opts)
  
print(optimalWeights.transpose())
## step 3: testing  
print("step 3: testing...")  
accuracy = testLogRegres(optimalWeights, test_x, test_y)  
  
## step 4: show the result  
print("step 4: show the result...")    
print('The classify accuracy is: %.3f%%' % (accuracy * 100)) 

step 2: training...
Error of input is: [[ 0.55020596]]

[[-0.67826324 -0.90828869  1.29891616 -1.54594599  0.09601133 -1.52143659
  -0.0340893   1.83221197 -1.11267587  0.68557871  2.05633956  0.45317084
  -1.38018909 -0.88556539  1.24783298  1.66839032  1.93862715 -1.86493214
   2.02435254 -0.83407311 -1.45814607]]
step 3: testing...
Error of output is: [[ 0.57758489]]

step 4: show the result...
The classify accuracy is: 81.733%


In [35]:
## step 2: training...  
print("step 2: training...")  
opts = {'alpha': 0.05, 'maxIter': 2333, 'optimizeType': 'gradDescent'}  
optimalWeights = trainLogRegres(train_x, train_y, opts)
  
print(optimalWeights.transpose())
## step 3: testing  
print("step 3: testing...")  
accuracy = testLogRegres(optimalWeights, test_x, test_y)  
  
## step 4: show the result  
print("step 4: show the result...")    
print('The classify accuracy is: %.3f%%' % (accuracy * 100)) 

step 2: training...
Error of input is: [[ 11.21713337]]

[[-32.01856896 -28.6596104   35.16036555 -48.09169144  -0.42625936
  -49.57613521  -4.48298079  46.76817399 -31.66092971  20.39356144
   57.89219242   9.19010902 -39.36255989 -28.8717769   33.38720749
   45.91434892  50.0175759  -58.33231406  57.47767261 -26.2554072
  -51.14191945]]
step 3: testing...
Error of output is: [[ 12.21447953]]

step 4: show the result...
The classify accuracy is: 63.300%


In [36]:
## step 2: training...  
print("step 2: training...")  
opts = {'alpha': 0.005, 'maxIter': 2333, 'optimizeType': 'gradDescent'}  
optimalWeights = trainLogRegres(train_x, train_y, opts)
  
print(optimalWeights.transpose())
## step 3: testing  
print("step 3: testing...")  
accuracy = testLogRegres(optimalWeights, test_x, test_y)  
  
## step 4: show the result  
print("step 4: show the result...")    
print('The classify accuracy is: %.3f%%' % (accuracy * 100)) 

step 2: training...
Error of input is: [[ 1.36162082]]

[[-3.1223357  -2.84517565  3.33529599 -4.59971463 -0.07943446 -4.59202993
  -0.37412498  4.63379717 -3.29247014  1.81411094  5.6514972   1.04387252
  -3.9047586  -2.85612157  3.25651612  4.46983198  5.16899679 -5.51047803
   5.55677417 -2.63272552 -4.59776347]]
step 3: testing...
Error of output is: [[ 1.46014497]]

step 4: show the result...
The classify accuracy is: 64.800%


## Essay Questions

_You may choose to do the essay questions here in the .ipynb notebook, but you are welcomed to use a word processor instead and write your solutions there instead (and convert it into .pdf format).  If you do that, please ensure to delete this section._

1\. [LFD Exercise 1.2] Suppose that we use a perceptron to detect spam messages. Let's say that each email messages represented by the frequency of occurrence of keywords, and the output is +1 if the message is considered spam.

    1. Can you think of some keywords that will end up with a large positive weight into perceptron?

    2. How about keywords that will get a negative weight?

    3. What parameter in the perceptron directly affects how many borderline messages end up classified as spam?


1. Keywords that lead to a positive weight indicates that messages with the keyword are more likely to be spamming. Some words that are very casual or oral , or leading receiver to pay money are likely to be spamming. For examples, 'pay','credit card' or 'buy'.

2. Keywords contribute a negative weight if they are not likely to be spamming. If words are very written or official, they are likely to be not spamming. Examples like 'purchase'.

3. Parameters to control the strictness of the checking are like message length or message sending rate.

2\. Consider a coin tossing experiment. You toss a coin 100 times, with the result of heads 70 times and tails 30 times. We denote the probability of heads of this coin as Θ. Now consider a coin toss.

    1. Build a model using maximum lilkelihood estimation (MLE) to infer Θ.

    2. Can we judge that this is an unfair coin? Explain your answer.

1. We can probably use logistic regression to do the estimation of classification. If head faces up, note the result as 1,otherwise 0. 
    A sample model is to use sigmod function = 1 / (1 + e^(-f(x))) in which f(x) is a linear regression with regards to the parameters like Throwing angle , starting speed, or air speed. Suppose the measurements of the parameters are possible, we can conduct logistic regression to gain the weightage for each attribute. 
2. We cannot conclude that this is an unfair coin, since we need to analyze the obtained model. The above mentioned attributes are likely to affect the final result. The higher chance of head tossing may be due to the person's throwing sense. We will nedd to analyze the obtained weightage and compare it with a normal coin to see whether this one is unfair or not.

3\. In the programming logistic regression, part (c), we did away with the stochastic idea of SGD and substituted a round-robin version, which deterministically uses the next point in turn to perform the gradient descent. Describe whether you think this is a good robust idea or not for datasets in general.

A round robin method may not be robust since it may be stuck in the local extreme value. If the function has both minimum and global minimum, a round robin method is likely to stop at a local minimum, which is not really optimal. There must be some methods to escape from local extreme value. SGD method can avoid local minimum by randomly jump through the local extrme value. Besides, since a deterministic method is used, the obtained result must be same every time to run it. This is not suitable to experiment for multiple different results.

## Statement of Individual Work

Please initial (between the square brackets) one of the following statements.

[I] I,A0153633J, certify that I have followed the CS 3244 Machine Learning class guidelines for homework assignments.  In particular, I expressly vow that I have followed the Facebook rule in discussing with others in doing the assignment and did not take notes (digital or printed) from the discussions.  

[ ] I, <*substitute your matric number here*>, did not follow the class rules regarding the homework assignment, because of the following reason:

<*Please fill in*>

I suggest that I should be graded as follows:

A

### References

I have refered to the following list of people and websites in preparing my homework submission:

http://docs.scipy.org/doc/numpy/reference/index.html
http://matplotlib.org/contents.html